### DISCUSSION TITLE
What networking library to use for a multiplayer game

### DISCUSSION THREAD

**Discussion by Tricked-dev** [bevyengine/bevy#4388] [Q&A]
Hello i am wanting to make a multiplayer game using bevy but i am running into the issue of what networking library to use.
The game i'm making must be able to have 100+ players in the same "lobby" and see each other move. i am considering just using normal web-socket's.

libraries i have tried:
- [naia](https://github.com/naia-rs/naia/tree/main/demos/bevy) cant get it to compile
- [bootleg_networking](https://github.com/billyb2/bootleg_networking) i don't like its style + it seems unmaintained
- [spicy_networking](https://github.com/CabbitStudios/bevy_spicy_networking) outdated ( does have a bevy 0.6 pr but not accepted)

I was wondering if any of you have a suggestion for a networking library to use or if i should go ws/tcp?

edit:
noticed that this was on top of search results and is still being actively viewed see <https://bevyengine.org/assets/#networking> for a list of libraries

**Comment by NiklasEi** [comment]
I found matchbox very nice. They have [an example with bevy and ggrs](https://github.com/johanhelsing/matchbox/tree/main/matchbox_demo).

For a game where you want to constantly sync positions/actions in real time I would not suggest websockets. They can be a lot easier for things like turn based games, but your use case sounds like WebRTC would fit nicely.

**Reply by Tricked-dev** [reply]
Thanks ill check it out

**Reply by minecrawler** [reply]
WebRTC means that clients are authoritative. That's not something you always want. Just think of all the cheating possibilities and no single server you control to check values.
Also, a lot of clients may be behind firewalls or proxies and will need an TURN server anyway.

**Reply by Tricked-dev** [reply]
Yeah that's the annoying part might just have to with ws then.

**Reply by james7132** [reply]
While the TURN server/NAT punch through point is indeed a common problem, the networking model of ggrs requires zero state transfer over the network. Only inputs are sent over the network. If no state is transferred, no authority is needed.

**Reply by LucCADORET** [reply]
WebRTC does not mean your game is client-authoritative. You can have an authoritative server broadcasting with WebRTC to the clients, acting as the only authority. You can use libraries like https://github.com/kyren/webrtc-unreliable to setup WebRTC server-side. What is your issue with Naia ? The bevy example works fine for me.

**Comment by heavyrain266** [comment]
I suggest building internal library with [quinn](https://github.com/quinn-rs/quinn) which is high level QUIC (UDP + TCP mix) protocol implementation which is quite easy to use.

**Comment by caniko** [comment]
[wtransport](https://github.com/BiagioFesta/wtransport) is new! It should even work on the web!

**Comment by NHodgesVFX** [comment]
Some more suggestions

https://github.com/lucaspoffo/renet
https://github.com/lifescapegame/bevy_replicon
https://github.com/gschup/bevy_ggrs

Tutorial for bevy_ggrs and matchbox

https://johanhelsing.studio/posts/extreme-bevy

**Comment by mogambro** [comment]
I'd use https://github.com/Henauxg/bevy_quinnet

**Comment by cBournhonesque** [comment]
You can also try 
https://github.com/cBournhonesque/lightyear 
which also does prediction/replication and is transport-agnostic
